src/toxcore/dht/server.rs

```diff
@@ -170,26 +179,29 @@ impl Server {
         match packet {
             DhtPacket::PingRequest(packet) => {
                 debug!("Received ping request");
-                let client = self.create_client(&addr, packet.pk);
-                self.handle_ping_req(client, packet)
+                self.handle_ping_req(packet, addr)
             },
             DhtPacket::PingResponse(packet) => {
                 debug!("Received ping response");
-                if let Some(client) = self.get_client(&packet.pk) {
-                    self.handle_ping_resp(client, packet)
+                let mut state = self.state.write();
+                if let Some(mut client) = self.get_client(&packet.pk, &state.peers_cache) {
+                    let result = self.handle_ping_resp(client.clone(), packet.clone(), &mut state.kbucket);
+                    client.last_resp_time = Instant::now();
+                    state.peers_cache.insert(packet.pk, client);
+                    result
                 } else { // If there doesn't exist client in hash, then the PingRequest sent is not from me. Do nothing.
                     Box::new(future::ok(()))
                 }
             },
             DhtPacket::NodesRequest(packet) => {
                 debug!("Received NodesRequest");
-                let client = self.create_client(&addr, packet.pk);
-                self.handle_nodes_req(client, packet)
+                self.handle_nodes_req(packet, addr)
             },
             DhtPacket::NodesResponse(packet) => {
                 debug!("Received NodesResponse");
-                if let Some(client) = self.get_client(&packet.pk) {
-                    self.handle_nodes_resp(client, packet)
+                let mut state = self.state.write();
+                if let Some(client) = self.get_client(&packet.pk, &state.peers_cache) {
```

Comment:
You must call `handle_*(packet, addr)` here, while you take locks inside handle_* function.